src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+    type: CycloneDxComponentType = attr.ib(
+        default=CycloneDxComponentType.LIBRARY)
+    properties: List[CycloneDxAttribute] = attr.ib(factory=list)
+    scope: CycloneDxComponentScope = attr.ib(default=CycloneDxComponentScope.REQUIRED)
+
+
+@attr.s
+class CycloneDxDependency:
+    ref: str = attr.ib()
+    dependsOn: str = attr.ib(factory=list)
+
+
+@attr.s
+class CycloneDxBom():
+    bomFormat: str = attr.ib(init=False, default="CycloneDX")
+    specVersion: str = attr.ib(init=False, default="1.3")
+    serialNumber: str = attr.ib(
+        init=False, default="urn:uuid:" + str(uuid.uuid4()))
+    version: int = attr.ib(init=False, default=1)
+    metadata: CycloneDxMetadata = attr.ib(default=CycloneDxMetadata())
+    components: List[CycloneDxComponent] = attr.ib(factory=list)
+    dependencies: List[CycloneDxDependency] = attr.ib(factory=list)
+
+
+def get_tool_header(version: str) -> dict:
+    return {
+        "vendor": "nexB Inc.",
+        "name": "scancode-toolkit",
+        "version": version
+    }
+
+
+def get_external_ref_from_key(package: dict, key: str) -> CycloneDxExternalRef:
+    url = package.get(key)
+    type = url_type_mapping[key]
+    if url is not None and type is not None:
+        return CycloneDxExternalRef(url=url, type=type)
+    return None
+
+
+def get_license_entry_from_license_expression(license_expression: str) \
+        -> CycloneDxLicenseEntry:
+    """query our list of known licenses to see if we can resolve
+    the license_expression to a single entry"""
+
+    # exit early if we don't have a valid license_expression
+    if license_expression is None:
+        return None
+
+    license = known_licenses.get(license_expression)
+    if license is not None:
+        # attempt to set either official OSI URL or any license text URL
+        url = license.osi_url
+        if license.text_urls:
+            url = license.text_urls[0]
+        lic = CycloneDxLicense(id=license.spdx_license_key,
+                               name=license.name,
+                               url=url)
+        return CycloneDxLicenseEntry(license=lic, expression=None)
+    else:
+        return CycloneDxLicenseEntry(license=None, expression=license_expression)
+
+
+url_pattern = re.compile(r"^(https?://)?[^\s/$.?#].[^\s]*$")
+
+
+def get_license_entry_from_declared_license(declared_license) \
+        -> CycloneDxLicenseEntry:
+    lic = CycloneDxLicense()
+    if isinstance(declared_license, str):
+        # if license key is in list of known spdx ids, set as id
+        if declared_license in spdx_ids:
+            lic.id = declared_license
+        # if we match this regex assume we are dealing with a URL
+        elif declared_license is not None and url_pattern.match(declared_license):
+            lic.url = declared_license
+        else:
+            lic.name = declared_license
+    # some declared_license entries are also expressed as dicts
```

Comment:
.... and some as list or anything. By construction the data structure is whatever existed upstream. I think you should not use the declared license in the output but instead only the `license_expression` which is the detection that took place